### PR TITLE
Java fix

### PR DESCRIPTION
**Pull request by cbielow**
We are constantly running into issues that TOPP tools cannot find Java, usually because App-bundles on MacOSX mess up the PATH, such that a simple call to 'java' fails.
This affects all TOPP tools internally calling Java, i.e. MSGF+ and Luciphor

This PR fixes this by:
 - allow path to Java to be specified explicitly
 - better Java checking and error messages

replaces #1941

### CODE REVIEW

**Comment by lars20070** [comment]
Resolves #1764 under OS X 10.11.5. An explicit java path is a clean solution. Thanks @cbielow!

+1

**Comment by jpfeuffer** [comment]
Should we remove these additional parameters for KNIME, since it will choose its own Java Runtime?
See
https://github.com/OpenMS/OpenMS/blob/develop/cmake/knime_package_support.cmake#L130

**Comment by hroest** [comment]
+1 @cbielow great work, especially printing out the PATH into the output should help to debug users problems

**Comment by cbielow** [comment]
The MSGF+ adapter tests fail with 
```
Java-Check:  Java not found at 'java'!
  Make sure Java is installed and this location is correct.
  You might need to add the Java binary to your PATH variable
  or use an absolute path+filename pointing to Java.
  The current SYSTEM PATH is: '/home/travis/build/OpenMS/OpenMS/_thirdparty/Fido:/home/travis/build/OpenMS/OpenMS/_thirdparty/MSGFPlus:/home/travis/build/OpenMS/OpenMS/_thirdparty/XTandem:/home/travis/build/OpenMS/OpenMS/_thirdparty/OMSSA:/home/travis/build/OpenMS/OpenMS/_thirdparty/MyriMatch:/home/travis/bin:/home/travis/.local/bin:/home/travis/.gimme/versions/go1.4.1.linux.amd64/bin:/usr/local/phantomjs/bin:/home/travis/.nvm/v0.10.36/bin:./node_modules/.bin:/usr/local/maven-3.2.5/bin:/usr/local/clang-3.4/bin:/home/travis/.gimme/versions/go1.4.1.linux.amd64/bin:/usr/local/phantomjs/bin:./node_modules/.bin:/usr/local/maven-3.2.5/bin:/usr/local/clang-3.4/bin:/home/travis/.gimme/versions/go1.4.1.linux.amd64/bin:/home/travis/.rvm/gems/ruby-1.9.3-p551/bin:/home/travis/.rvm/gems/ruby-1.9.3-p551@global/bin:/home/travis/.rvm/rubies/ruby-1.9.3-p551/bin:/usr/local/phantomjs/bin:./node_modules/.bin:/usr/local/maven-3.2.5/bin:/usr/local/clang-3.4/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/travis/.rvm/bin:/home/travis/.rvm/bin:/home/travis/.rvm/bin'.


Fatal error: Java is needed to run MS-GF+!
MSGFPlusAdapter took 0.00 s (wall), 0.00 s (CPU), 0.00 s (system), 0.00 s (user).
```

This is weird, since the calling string is still "java". I currently at a loss why this happens....

**Comment by jpfeuffer** [comment]
Well, how was it called? The standard value for the parameter is set to "java"

File: src/openms/source/SYSTEM/JavaInfo.cpp [hunk 1]
```diff
@@ -33,17 +33,72 @@
...
+          static String path;
+          if (path.empty())
+          {
+            path = getenv("PATH");
+          }
+          LOG_ERROR << "  You might need to add the Java binary to your PATH variable\n"
+                    << "  or use an absolute path+filename pointing to Java.\n" 
+                    << "  The current SYSTEM PATH is: '" << path << "'.\n\n"
+#ifdef __APPLE__
+                    << "  On MacOSX, application bundles change the system PATH; Use an abolute path to Java or open your exectuable (e.g. KNIME/TOPPAS/TOPPView) from within the bundle!\n"
```
Comment (jpfeuffer) [inline src/openms/source/SYSTEM/JavaInfo.cpp]: absolute, executable

**Comment by cbielow** [comment]
the invocation is (should be) exactly the same as before with the 
current defaults, i.e. "java".



Am 7/25/2016 um 11:35 AM schrieb jpfeuffer:

**Comment by jpfeuffer** [comment]
Ah I see what you mean, the call was not changed. Sorry.
It seems that our File::find() just does not work.
qp.start seems to find it even without this check. Maybe related to all the OpenMS.ini, OpenMS/share problems we have.

**Comment by cbielow** [comment]
I've played around with File::find() to teach it respecting the system PATH, but then there are other obstacles under Windows, i.e. passing 'java' to QFileInfo::exists() will return false, since it should be 'java.exe'. Also quoting the string '"java.exe"' fails.
QProcess cares about none of this and just works, so I've restricted the diagnosis to what QProcess provides -- its more then enough I think.

Lets see what the tests say,

**Comment by cbielow** [comment]
contrib used in Travis seems to be broken atm

**Comment by jpfeuffer** [comment]
https://github.com/OpenMS/contrib/pull/29
My bad.

**Comment by jpfeuffer** [comment]
Restarted travis.

**Comment by cbielow** [comment]
this seems to work.
Merge?

**Comment by jpfeuffer** [comment]
What do you think about removing it for KNIME. KNIME is always shipped with a current JRE (since it also runs in one), currently we set the internal PATH to this java executable.
Do you think users still want to have a choice to change it? Or does it confuse them to have that option?

Could be changed by removing it from the descriptors like this:
https://github.com/OpenMS/OpenMS/blob/develop/cmake/knime_package_support.cmake#L130

**Comment by cbielow** [comment]
mhh.. I would not hide it the param from the user. The default will be 
fine for most cases.

Maybe at some point MSGF+ needs a newer or older Java version than 
KNIME, and all of a sudden you'd want the switch back.



Am 7/28/2016 um 3:16 PM schrieb jpfeuffer:

**Comment by jpfeuffer** [comment]
Okay. As long as the default is "java" and it will find the executable on all platforms when the java directory is in the PATH (which it seems to do), this should be fine with KNIME also.
+1

**Pull request merged**